### ISSUE TITLE
Strict Header Binding issue

### ISSUE THREAD

**Issue by olibiaz** [eMAGTechLabs/RabbitMqBundle#30]
It seems impossible to use a strict `headers` exchange routing at the moment.

I basically want to be able to able to bind the queue `my-test-queue` to exchange `my-test-exchange` only via the header `x-test-header: ok`

currently i tried this configuration:

    producers:
       test-producer:
         class: My\Custom\Class
         connection: default
         auto_setup_fabric: false
         exchange_options: {name: my-test-exchange, type: headers}
         service_alias:  producer.test
         queue_options:
           name: my-test-queue
    consumers:
       test-consumer:
         connection: default
         auto_setup_fabric: false
         exchange_options: {name:  my-test-exchange, type: headers}
         callback:   consumer.weather_collector
    bindings:
      - exchange: my-test-exchange
        destination: my-test-queue
        routing_key: '*'
        arguments:
          'x-test-header': ['S', ok]
          'x-match': ['S', all]

The problem is, even with `bindings` option, it will have "two bindings" applied, the first one when configuration of `producers` is applied, then the second when the configuration of `bindings` is applied.

At the end i got this binding:

![Screenshot from 2020-08-26 13-33-05](https://user-images.githubusercontent.com/13057911/91337175-f9cd8200-e7a0-11ea-8f44-61245ea18868.png)

The "empty" binding is the problem here, since it allows me to send message to the queue via the exchange without the required header.

Do i miss something here? I tried several configurations without any success. Please let me know if there already has a way to achieve that.

If its not the case, I think there is several solutions to achieve a strict headers binding:

**Solution 1:** 

Allow to pass bind arguments at the queue configuration level. 

Currently, we can pass `arguments` to queue configuration, but this configuration is not forwarded to the `queueBind` method call in `BaseAmqp.queueDeclare()`

https://github.com/eMAGTechLabs/RabbitMqBundle/blob/d34844a0dc4d5f8ec5ba58a038a1d18e3e04856a/RabbitMq/BaseAmqp.php#L218-L236

https://github.com/eMAGTechLabs/RabbitMqBundle/blob/d34844a0dc4d5f8ec5ba58a038a1d18e3e04856a/RabbitMq/BaseAmqp.php#L245-L251

that use the `AMQPChanel.queue_bind()` (php-amqplib) method, that can handle `arguments`

    public function queue_bind(
        $queue,
        $exchange,
        $routing_key = '',
        $nowait = false,
        $arguments = array(),
        $ticket = null
    )

By forwarding the `arguments` configuration, we could directly insure a proper binding on the queue configuration level.

**Solution 2:** 

Unbinding empty default binding for source and destination when `Binding.setupFabric()` is called. This way, if the developer let the default empty the binding in the queue configuration like in my example:
    
    queue_options:
           name: my-test-queue # no binding here, so currently it will automatically bind it to routing_key: ''

It could unbind this automatically between `my-test-queue` and `my-test-exchange` here:

https://github.com/eMAGTechLabs/RabbitMqBundle/blob/d34844a0dc4d5f8ec5ba58a038a1d18e3e04856a/RabbitMq/Binding.php#L139-L151

For example by doing something like that (just an example):

    public function setupFabric()
    {
        $method = ($this->destinationIsExchange) ? 'exchange_bind' : 'queue_bind';
        $channel = $this->getChannel();

        if ($this->destinationIsExchange) {
            $channel->exchange_unbind($this->destination, $this->exchange);
        } else {
            $channel->queue_unbind($this->destination, $this->exchange);
        }

        call_user_func(
            array($channel, $method),
            $this->destination,
            $this->exchange,
            $this->routingKey,
            $this->nowait,
            $this->arguments
        );
    }

**Solution 3** 

Ignoring or overriding the queues declaration bindings for specific queues/exchanges if a `bindings` configuration is already set for these queues/exchanges.

Basically, if we set a `bindings` configuration that setup a binding between `exchange-a` and `queue-a`, I don't see the point to also apply the binding here:

https://github.com/eMAGTechLabs/RabbitMqBundle/blob/d34844a0dc4d5f8ec5ba58a038a1d18e3e04856a/RabbitMq/BaseAmqp.php#L226-L232

when doing the `queueDeclare()`.

Don't know what you guys think about that, but i would like to read your opinion on that.

thanks.

**Comment by mihaileu** [comment]
Hi @olibiaz, thanks for your proposal, we will look into to see what can we do.

**Comment by olibiaz** [comment]
Thank you @mihaileu,

I realized after writing this issue that the `x-test-header` naming i picked for my example may be misleading/error-prone, since headers exchanges ignore headers arguments prefixed by `x-` for routing. Better use another naming if you try reproduce the error.

**Comment by mihaileu** [comment]
2.2.2 tag was released
